### PR TITLE
Batching handled directly on a GPJax Dataset

### PR DESCRIPTION
**Pull request by Daniel-Dodd**
This PR puts mini-batching directly on the GPJax Dataset class #44. 

It utilises tensorflow.data as a backend. 

It currently supports:

- Batching
- Prefetching
- Shuffling
- Repeating
- Caching

The function 'fit_batches' from abstractions.py, now directly works on passing through a GPJax dataset, rather than constructing a mini-batcher.

Unit tests have been edited to adopt the new proposed framework, but no new ones have been created.

Screen shot of new syntax is attached.
Screenshot 2022-05-10 at 13 48 54 (2)

### CODE REVIEW

**Comment by Daniel-Dodd** [comment]
Documentation needs to be rewritten to accommodate changes.

File: gpjax/types.py [hunk 1]
```diff
@@ -31,6 +34,46 @@ def in_dim(self) -> int:
...
+        self._make_tfd_if_none()
+        self._tf_dataset = self._tf_dataset.batch(*args, **kwargs)
+        return self
+    
+    def prefetch(self, *args, **kwargs) -> "Dataset":
+        self._make_tfd_if_none()
+        self._tf_dataset = self._tf_dataset.prefetch(*args, **kwargs)
+        return self
+    
+    def get_batches(self) -> tp.Callable[[None], "Dataset"]:
```
Comment (thomaspinder) [inline gpjax/types.py:66]: Should this be called something like `make_batch_fn()`, `get_batch_fn()`, or `get_batcher()`. The current `get_batches()` name makes me think I'm going to receive a set of batches.

File: gpjax/types.py [hunk 2]
```diff
@@ -31,6 +34,46 @@ def in_dim(self) -> int:
     def out_dim(self) -> int:
         return self.y.shape[1]
 
+    def _make_tfd_if_none(self) -> None:
+        if self._tf_dataset is None:
+            self._tf_dataset = tfd.Dataset.from_tensor_slices((self.X, self.y))
+    
+    def cache(self, *args, **kwargs) -> "Dataset":
```
Comment (thomaspinder) [inline gpjax/types.py:41]: Can we just add a short, single sentence docstring to each of these methods please.

File: gpjax/types.py [hunk 3]
```diff
@@ -7,17 +7,20 @@
...
 
-@dataclass(repr=False)
+
+@dataclass
 class Dataset:
+    """GPJax Dataset."""
+
     X: Array
     y: Array = None
+    _tf_dataset: tp.Optional[tfd.Dataset] = None
```
Comment (thomaspinder) [inline gpjax/types.py:19]: Perhaps acknowledge what this is in the object's docstring e.g., line 15. It's not immediately clear

Comment (Daniel-Dodd) [reply gpjax/types.py:18]: Absolutely -> see https://github.com/thomaspinder/GPJax/pull/48/commits/64364a4fa3fbbbec39262f733017a4af6edd1648.

File: gpjax/types.py (hunk 1 above)
Comment (Daniel-Dodd) [reply gpjax/types.py]: Absolutely -> renamed to 'get_batcher' in [254643d](https://github.com/thomaspinder/GPJax/pull/48/commits/254643d4c32cb8a899d4395bd1b2055dfc5648e4).

**Pull request merged**